src/nvim/popupmenu.c

```diff
@@ -1491,3 +1523,131 @@ void pum_make_popup(const char *path_name, int use_mouse_pos)
     pum_show_popupmenu(menu);
   }
 }
+
+static bool pum_parse_title(WinConfig *fconfig, BorderTextType bt, const char *s, size_t len)
+{
+  Error err = ERROR_INIT;
+  bool result = true;
+  char *data = xmemdupz(s, len);
+  if (data == NULL) {
+    return false;
+  }
+  Object obj = OBJECT_INIT;
+  obj.type = kObjectTypeString;
+  obj.data.string.data = data;
+  obj.data.string.size = len;
+  parse_bordertext(obj, bt, fconfig, &err);
+  api_free_object(obj);
+  if (ERROR_SET(&err)) {
+    emsg(err.msg);
+    api_clear_error(&err);
+    result = false;
+  }
+  return result;
+}
+
+static bool pum_parse_title_pos(WinConfig *fconfig, BorderTextType bt, const char *s, size_t len)
+{
+  Error err = ERROR_INIT;
+  char *data = xmemdupz(s, len);
+  if (data == NULL) {
+    return false;
+  }
+  String str = {
+    .data = data,
+    .size = len,
+  };
+  bool result = parse_bordertext_pos(str, bt, fconfig, &err);
+  api_free_string(str);
+  if (ERROR_SET(&err)) {
+    emsg(err.msg);
+    api_clear_error(&err);
+    result = false;
+  }
+  return result;
+}
+
+bool parse_completepopup(WinConfig *fconfig, int *border_adj)
+{
+  const char *p = p_cpp;
+  Error err = ERROR_INIT;
+
+  while (*p != NUL) {
+    const char *s = p;
+
+    const char *e = vim_strchr(p, ':');
+    if (e == NULL || e[1] == NUL) {
+      return FAIL;
+    }
+
+    p = vim_strchr(e, ',');
+    if (p == NULL) {
+      p = e + strlen(e);
+    }
+
+    size_t len = (size_t)(p - e - 1);
+    if (strncmp(s, "border:", 7) == 0) {
+      Object style = OBJECT_INIT;
+      style.type = kObjectTypeString;
+      style.data.string.data = xmemdupz(e + 1, len);
+      if (!style.data.string.data) {
+        return false;
+      }
+      style.data.string.size = len;
+      parse_border_style(style, fconfig, &err);
+      if (ERROR_SET(&err)) {
+        emsg(err.msg);
+        api_clear_error(&err);
+        api_free_object(style);
+        return false;
+      }
+
+      int border_attr = win_hl_attr(curwin, HLF_BORDER);
+      for (int i = 0; i < 8; i++) {
+        fconfig->border_attr[i] = fconfig->border_hl_ids[i]
+                                  ? hl_get_ui_attr(0, HLF_BORDER, fconfig->border_hl_ids[i], false)
+                                  : border_attr;
+        if (i < 4 && fconfig->border_chars[2 * i + 1][0]) {
+          border_adj[i] = 1;
+        }
+      }
+
+      api_free_object(style);
+    } else if (strncmp(s, "title:", 6) == 0) {
+      if (!pum_parse_title(fconfig, kBorderTextTitle, e + 1, len)) {
+        return false;
+      }
+      fconfig->title = true;
+    } else if (strncmp(s, "title_pos:", 10) == 0) {
```

Comment:
A bit of a nit, but for most other options (like 'completeopt') we don't use underscores, so I think this should be `titlepos` (and similarly for `footerpos`).

Reply:
make sense.